### PR TITLE
homeScreen (favourite)

### PR DESCRIPTION
**Pull request by nhanviet1**
## Related Tickets
- [#Task 55197: Content](https://edu-redmine.sun-asterisk.vn/issues/55197)
## WHAT
- UI for Favourite Screen and image detail screen
- Add functions: Add, delete to the favourite. Download images to device
## Evidence (Screenshot or Video)
image
image
image

## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/WtYPAy |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

**Review by hungld-1302 — APPROVED** [review]

File: app/src/main/java/com/example/catapp/data/source/remote/fetchjson/ParseFavouriteJson.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package com.example.catapp.data.source.remote.fetchjson
+
+import com.example.catapp.data.model.responsemodel.CatEntry
+import com.example.catapp.data.model.responsemodel.breeds.Image
+import com.example.catapp.data.model.responsemodel.favourite.FavouriteItem
+import org.json.JSONObject
+
+class ParseFavouriteJson {
+
+    companion object {
```
Comment (hungvv-1301) [inline app/src/main/java/com/example/catapp/data/source/remote/fetchjson/ParseFavouriteJson.kt]: companion để cuối

File: app/src/main/java/com/example/catapp/data/source/remote/fetchjson/post/ConvertFavouriteJson.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.example.catapp.data.source.remote.fetchjson.post
+
+import org.json.JSONObject
+
+class ConvertFavouriteJson {
+
+    fun convertToJson(imgID: String): String {
+        val jsonObject = JSONObject().run {
+            put("image_id", imgID)
```
Comment (hungvv-1301) [inline app/src/main/java/com/example/catapp/data/source/remote/fetchjson/post/ConvertFavouriteJson.kt]: constant

File: app/src/main/java/com/example/catapp/data/source/remote/fetchjson/delete/ParseDeleteResultJson.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+import com.example.catapp.data.model.responsemodel.favourite.FavouriteResponse
+import com.example.catapp.utils.NONE
+import org.json.JSONObject
+
+class ParseDeleteResultJson {
+
+    fun delResultParse(jsonObject: JSONObject): FavouriteResponse {
+        val message = jsonObject.optString(responseMessage)
+
```
Comment (hungvv-1301) [inline app/src/main/java/com/example/catapp/data/source/remote/fetchjson/delete/ParseDeleteResultJson.kt]: space

File: app/src/main/java/com/example/catapp/utils/ImageDownloadUtil.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        while (downloading) {
+            val cursor: Cursor = downloadManager.query(query)
+            cursor.moveToFirst()
+            val status = cursor.getInt(
+                cursor.getColumnIndex(DownloadManager.COLUMN_STATUS).toInt()
+            )
+            if (status == DownloadManager.STATUS_SUCCESSFUL) {
+                downloading = false
+                mHandler.post {
+                    shortToast("Downloaded ($imgTitle) Successfully")
```
Comment (hungvv-1301) [inline app/src/main/java/com/example/catapp/utils/ImageDownloadUtil.kt]: .

File: app/src/main/java/com/example/catapp/view/homescreen/BigImageActivity.kt [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        deleteButton?.setOnClickListener {
+            if (userAPI != null && imgID != null) {
+                singleFavPresenter?.deleteRemoteFav(userAPI!!, imgID!!)
+                dialog?.dismiss()
+            }
+        }
+
+        favButton?.setOnClickListener {
+            if (urlList.contains(imgURL) && imgURL != null) {
+                this@BigImageActivity.shortToast("This image is already in your Favourite")
```
Comment (hungvv-1301) [inline app/src/main/java/com/example/catapp/view/homescreen/BigImageActivity.kt]: .

File: app/src/main/java/com/example/catapp/view/homescreen/BigImageActivity.kt [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+    override fun onPostFavSuccess(favResponse: FavouriteResponse) {
+        this@BigImageActivity.shortToast(favResponse.message)
+        Intent(RELOAD_FAV).run {
+            sendBroadcast(this)
+        }
+    }
+
+    override fun onDeleteFavSuccess(favResponse: FavouriteResponse) {
+        this@BigImageActivity.shortToast("Deleted Successfully")
```
Comment (hungvv-1301) [inline app/src/main/java/com/example/catapp/view/homescreen/BigImageActivity.kt]: .

File: app/src/main/java/com/example/catapp/view/loginscreen/LoginActivity.kt [hunk 7]
```diff
@@ -33,7 +33,6 @@ class LoginActivity : BaseActivity<ActivityLoginBinding>(ActivityLoginBinding::i
         val userAPI = binding.editTextRequestAPI.text.toString()
         val checker = cat[0].breed
         if (checker == null) {
-            startHomeScreen(userAPI)
             shortToast("Your API is invalid!")
```
Comment (hungvv-1301) [inline app/src/main/java/com/example/catapp/view/loginscreen/LoginActivity.kt]: .

File: app/src/main/java/com/example/catapp/view/homescreen/homefragments/FavouriteFragment.kt [hunk 8]
```diff
@@ -18,5 +41,59 @@ class FavouriteFragment : Fragment() {
...
+            context?.registerReceiver(favReceiver, intentFilter)
+            isRegistered = true
+        }
+    }
+
+    override fun onDetach() {
+        super.onDetach()
+        if (isRegistered) {
+            context?.unregisterReceiver(favReceiver)
+        }
```
Comment (hungvv-1301) [inline app/src/main/java/com/example/catapp/view/homescreen/homefragments/FavouriteFragment.kt:96]: cân nhắc

**Review by ndt0409 — APPROVED** [review]

File: app/src/main/java/com/example/catapp/data/source/remote/fetchjson/GetJson.kt [hunk 9]
```diff
@@ -87,5 +88,7 @@ class GetJson<T>(
         private const val TIME_OUT = 15000
         private const val METHOD_GET = "GET"
         private const val X_API_KEY = "x-api-key"
+        private const val CONTENT_TYPE_HEAD = "Content-Type"
+        private const val CONTENT_TYPE = "application/json"
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/catapp/data/source/remote/fetchjson/GetJson.kt]: buildconfigfield đi em

File: app/src/main/java/com/example/catapp/data/source/remote/fetchjson/delete/DeleteJson.kt [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        }
+        bufferedReader.close()
+        httpURLConnection?.disconnect()
+        return stringBuilder.toString()
+    }
+
+    companion object {
+        private const val TIME_OUT = 15000
+        private const val METHOD_DELETE = "DELETE"
+        private const val X_API_KEY = "x-api-key"
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/catapp/data/source/remote/fetchjson/delete/DeleteJson.kt]: khai báo lặp này

File: app/src/main/java/com/example/catapp/view/homescreen/BigImageActivity.kt [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+        dialog?.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
+        dialog?.window?.attributes?.windowAnimations = R.style.DialogAnimation
+        dialog?.window?.setGravity(Gravity.BOTTOM)
+        dialogButtonSetup()
+    }
+
+    private fun dialogButtonSetup() {
+        val favButton = dialog?.findViewById<TextView>(R.id.text_add_to_fav)
+        val downloadButton = dialog?.findViewById<TextView>(R.id.text_download)
+        val deleteButton = dialog?.findViewById<TextView>(R.id.text_delete)
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/catapp/view/homescreen/BigImageActivity.kt]: viewbinding mà em

File: app/src/main/java/com/example/catapp/view/homescreen/BigImageActivity.kt [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<out String>,
+        grantResults: IntArray
+    ) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        if (requestCode == REQUEST_CODE && grantResults[NONE] == PackageManager.PERMISSION_GRANTED) {
+            if (imgURL != null) {
+                downloadImg(imgURL!!)
+            }
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/catapp/view/homescreen/BigImageActivity.kt:175]: cái này để làm gì thế em , em handle permission granted bên trên r mà

File: app/src/main/java/com/example/catapp/view/homescreen/favpresenter/FavPresenter.kt [hunk 13]
```diff
@@ -0,0 +1,24 @@
+package com.example.catapp.view.homescreen.favpresenter
+
+import com.example.catapp.data.model.responsemodel.favourite.FavouriteItem
+import com.example.catapp.data.source.repository.FavRepository
+import com.sun.mvp.data.repository.source.remote.OnResultListener
+import java.lang.Exception
+
+class FavPresenter(
+    private val favRepository: FavRepository, private val view: FavInterface.View
+) : FavInterface.Presenter {
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/catapp/view/homescreen/favpresenter/FavPresenter.kt]:
```suggestion
class FavPresenter(
    private val favRepository: FavRepository,
    private val view: FavInterface.View
) : FavInterface.Presenter {
```

File: app/src/main/java/com/example/catapp/view/homescreen/favpresenter/singlefavpresenter/SingleFavPresenter.kt [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+import com.example.catapp.data.model.responsemodel.favourite.FavouriteItem
+import com.example.catapp.data.model.responsemodel.favourite.FavouriteResponse
+import com.example.catapp.data.source.repository.SingleFavRepository
+import com.sun.mvp.data.repository.source.remote.OnResultListener
+import java.lang.Exception
+
+class SingleFavPresenter(
+    private val favRepository: SingleFavRepository, private val view: SingleFavInterface.View
+) : SingleFavInterface.Presenter {
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/catapp/view/homescreen/favpresenter/singlefavpresenter/SingleFavPresenter.kt]:
```suggestion
class SingleFavPresenter(
    private val favRepository: SingleFavRepository,
    private val view: SingleFavInterface.View
) : SingleFavInterface.Presenter {
```

File: app/src/main/res/anim/slide_in_bottom.xml [hunk 15]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<set xmlns:android="http://schemas.android.com/apk/res/android">
+    <translate
+        android:duration="@android:integer/config_mediumAnimTime"
+        android:fromYDelta="100%p"
+        android:toYDelta="0"
+        />
```
Comment (vietnh-1109) [inline app/src/main/res/anim/slide_in_bottom.xml]:
```suggestion
        android:toYDelta="0" />
```

**Pull request merged**